File: leetcode/trie/211. Add and Search Word - Data structure design/js/main.js
// addWord () - O(N), N = длина нового слова
// search () - Наихудший случай: O(M), M = общее количество символов в Trie
class WordDictionary {
  constructor() {
    this.root = new Node();
  }

  addWord(word) {
    let node = this.root;

    for (let w of word) {
      if (!(w in node.children)) {
        node.children[w] = new Node();
      }
      node = node.children[w];
    }

    node.isWord = true;
  }

  dfs(word, node, index) {
    if (index >= word.length) {
      return node.isWord;
    }

    if (word[index] == '.') {
      for (let i = 0; i < 26; i++) {
        let ch = String.fromCharCode(i + 97);

        if (ch in node.children && this.dfs(word, node.children[ch], index + 1)) {
          return true;
        }
      }
      return false;
    }

    if (!(word[index] in node.children)) {
      return false;
    }

    return this.dfs(word, node.children[word[index]], index + 1);
  }

  search(word) {
    return this.dfs(word, this.root, 0);
  }
}

class Node {
  constructor() {
    this.isWord = false;
    this.children = {};
    this.word = null;
  }
}

File: leetcode/trie/211. Add and Search Word - Data structure design/py/main.py
class WordDictionary:

    def __init__(self):
        self.tree = Trie()        

    def addWord(self, word: str) -> None:
        self.tree.insert(word)   

    def search(self, word: str) -> bool:
        return self.tree.search(word)

class Trie:

    def __init__(self):
        self.root = Node()        

    def insert(self, word: str) -> None:
        node = self.root
        
        for w in word:
            if w not in node.children:
                node.children[w] = Node()
            node = node.children[w]        
        
        node.isWord = True

    def search(self, word: str) -> bool:
        n = len(word)

        def dfs(index, current):
          if index >= n:
            return current.isWord
          
          if word[index] == '.':
            for k in current.children.keys():
              if dfs(index + 1, current.children[k]):
                return True

          if word[index] in current.children and dfs(index + 1, current.children[word[index]]):
            return True
          
          return False

        return dfs(0, self.root)

class Node:

    def __init__(self, ):
      self.children = {}
      self.isWord = False
